`make reformat` fails (I'm using Python 3.7)

$ make reformat
# Create our Python 3.6 virtual environment
rm -rf .state/env
python3.6 -m venv .state/env
make: python3.6: No such file or directory
make: *** [.state/env/pyvenv.cfg] Error 1

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -25,7 +25,7 @@ jobs:
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Dependencies
-            command: bin/deps
+            command: bin/github-actions-deps
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Licenses

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -33,7 +33,7 @@ jobs:
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Translations
-            command: make translations BINDIR="$(dirname $(which python))"
+            command: bin/translations
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Static Tests

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -25,7 +25,7 @@ jobs:
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Dependencies
-            command: make github-actions-deps BINDIR="$(dirname $(which python))"
+            command: bin/deps
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Licenses

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -29,7 +29,7 @@ jobs:
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Licenses
-            command: make licenses BINDIR="$(dirname $(which python))"
+            command: bin/licenses
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Translations

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -21,7 +21,7 @@ jobs:
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Documentation
-            command: make docs BINDIR="$(dirname $(which python))"
+            command: bin/docs
             needs-python: ${{ true }}
             needs-node: ${{ false }}
           - name: Dependencies

--- a/Dockerfile
+++ b/Dockerfile
@@ -103,7 +103,7 @@ RUN set -x \
             install --no-deps --no-binary hiredis \
                     -r /tmp/requirements/deploy.txt \
                     -r /tmp/requirements/main.txt \
-                    $(if [ "$DEVEL" = "yes" ]; then echo '-r /tmp/requirements/tests.txt -r /tmp/requirements/lint.txt'; fi) \
+                    $(if [ "$DEVEL" = "yes" ]; then echo '-r /tmp/requirements/tests.txt -r /tmp/requirements/lint.txt -r /tmp/requirements/docs.txt'; fi) \
     && find /opt/warehouse -name '*.pyc' -delete
 
 
@@ -138,7 +138,7 @@ RUN set -x \
     && apt-get update \
     && apt-get install --no-install-recommends -y \
         libpq5 libxml2 libxslt1.1 libcurl4  \
-        $(if [ "$DEVEL" = "yes" ]; then echo 'bash libjpeg62 postgresql-client'; fi) \
+        $(if [ "$DEVEL" = "yes" ]; then echo 'build-essential bash libjpeg62 postgresql-client'; fi) \
     && apt-get clean \
     && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 

--- a/Dockerfile
+++ b/Dockerfile
@@ -138,7 +138,7 @@ RUN set -x \
     && apt-get update \
     && apt-get install --no-install-recommends -y \
         libpq5 libxml2 libxslt1.1 libcurl4  \
-        $(if [ "$DEVEL" = "yes" ]; then echo 'build-essential bash libjpeg62 postgresql-client'; fi) \
+        $(if [ "$DEVEL" = "yes" ]; then echo 'bash libjpeg62 postgresql-client build-essential libffi-dev libxml2-dev libxslt-dev libpq-dev libcurl4-openssl-dev libssl-dev'; fi) \
     && apt-get clean \
     && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 

--- a/Makefile
+++ b/Makefile
@@ -84,15 +84,6 @@ requirements/%.txt: requirements/%.in
 	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
 	  bin/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
 
-github-actions-deps:
-ifneq ($(GITHUB_BASE_REF), false)
-	git fetch origin $(GITHUB_BASE_REF):refs/remotes/origin/$(GITHUB_BASE_REF)
-	# Check that the following diff will exit with 0 or 1
-	git diff --name-only FETCH_HEAD || test $? -le 1 || exit 1
-	# Make the dependencies if any changed files are requirements files, otherwise exit
-	git diff --name-only FETCH_HEAD | grep '^requirements/' || exit 0 && bin/deps
-endif
-
 initdb:
 	docker-compose run --rm web psql -h db -d postgres -U postgres -c "SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE datname ='warehouse';"
 	docker-compose run --rm web psql -h db -d postgres -U postgres -c "DROP DATABASE IF EXISTS warehouse"

--- a/Makefile
+++ b/Makefile
@@ -165,40 +165,11 @@ purge: stop clean
 stop:
 	docker-compose down -v
 
-compile-pot: .state/env/pyvenv.cfg
-	PYTHONPATH=$(PWD) $(BINDIR)/pybabel extract \
-		-F babel.cfg \
-		--omit-header \
-		--output="warehouse/locale/messages.pot" \
-		warehouse
+translations: .state/docker-build
+	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
+								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+								  bin/translations
 
-init-po: .state/env/pyvenv.cfg
-	$(BINDIR)/pybabel init \
-		--input-file="warehouse/locale/messages.pot" \
-		--output-dir="warehouse/locale/" \
-		--locale="$(L)"
 
-update-po: .state/env/pyvenv.cfg
-	$(BINDIR)/pybabel update \
-		--input-file="warehouse/locale/messages.pot" \
-		--output-file="warehouse/locale/$(L)/LC_MESSAGES/messages.po" \
-		--locale="$(L)"
-
-compile-po: .state/env/pyvenv.cfg
-	$(BINDIR)/pybabel compile \
-		--input-file="warehouse/locale/$(L)/LC_MESSAGES/messages.po" \
-		--directory="warehouse/locale/" \
-		--locale="$(L)"
-
-build-mos: compile-pot
-	for LOCALE in $(LOCALES) ; do \
-		L=$$LOCALE $(MAKE) compile-po ; \
-		done
-
-translations: compile-pot
-ifneq ($(GITHUB_ACTIONS), false)
-	git diff --quiet ./warehouse/locale/messages.pot || (echo "There are outstanding translations, run 'make translations' and commit the changes."; exit 1)
-else
-endif
 
 .PHONY: default build serve initdb shell tests docs deps clean purge debug stop compile-pot

--- a/Makefile
+++ b/Makefile
@@ -1,38 +1,13 @@
 BINDIR = $(PWD)/.state/env/bin
-GITHUB_ACTIONS := $(shell echo "$${GITHUB_ACTIONS:-false}")
 GITHUB_BASE_REF := $(shell echo "$${GITHUB_BASE_REF:-false}")
 DB := example
 IPYTHON := no
-LOCALES := $(shell .state/env/bin/python -c "from warehouse.i18n import KNOWN_LOCALES; print(' '.join(set(KNOWN_LOCALES)-{'en'}))")
 
 # set environment variable WAREHOUSE_IPYTHON_SHELL=1 if IPython
 # needed in development environment
 ifeq ($(WAREHOUSE_IPYTHON_SHELL), 1)
     IPYTHON = yes
 endif
-
-define DEPCHECKER
-import sys
-from pip_api import parse_requirements
-
-left, right = sys.argv[1:3]
-left_reqs = parse_requirements(left).keys()
-right_reqs = parse_requirements(right).keys()
-
-extra_in_left = left_reqs - right_reqs
-extra_in_right = right_reqs - left_reqs
-
-if extra_in_left:
-	for dep in sorted(extra_in_left):
-		print("- {}".format(dep))
-
-if extra_in_right:
-	for dep in sorted(extra_in_right):
-		print("+ {}".format(dep))
-
-if extra_in_left or extra_in_right:
-	sys.exit(1)
-endef
 
 default:
 	@echo "Call a specific subcommand:"
@@ -116,17 +91,10 @@ docs: .state/docker-build
 licenses:
 	bin/licenses
 
-export DEPCHECKER
-deps: .state/env/pyvenv.cfg
-	$(eval TMPDIR := $(shell mktemp -d))
-	$(BINDIR)/pip-compile --upgrade --allow-unsafe -o $(TMPDIR)/deploy.txt requirements/deploy.in > /dev/null
-	$(BINDIR)/pip-compile --upgrade --allow-unsafe -o $(TMPDIR)/main.txt requirements/main.in > /dev/null
-	$(BINDIR)/pip-compile --upgrade --allow-unsafe -o $(TMPDIR)/lint.txt requirements/lint.in > /dev/null
-	echo "$$DEPCHECKER" | $(BINDIR)/python - $(TMPDIR)/deploy.txt requirements/deploy.txt
-	echo "$$DEPCHECKER" | $(BINDIR)/python - $(TMPDIR)/main.txt requirements/main.txt
-	echo "$$DEPCHECKER" | $(BINDIR)/python - $(TMPDIR)/lint.txt requirements/lint.txt
-	rm -r $(TMPDIR)
-	$(BINDIR)/pip check
+deps: .state/docker-build
+	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
+								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+								  bin/deps
 
 requirements/%.txt: requirements/%.in .state/env/pyvenv.cfg
 	$(BINDIR)/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
@@ -137,7 +105,7 @@ ifneq ($(GITHUB_BASE_REF), false)
 	# Check that the following diff will exit with 0 or 1
 	git diff --name-only FETCH_HEAD || test $? -le 1 || exit 1
 	# Make the dependencies if any changed files are requirements files, otherwise exit
-	git diff --name-only FETCH_HEAD | grep '^requirements/' || exit 0 && $(MAKE) deps
+	git diff --name-only FETCH_HEAD | grep '^requirements/' || exit 0 && bin/deps
 endif
 
 initdb:

--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,3 @@
-BINDIR = $(PWD)/.state/env/bin
 GITHUB_BASE_REF := $(shell echo "$${GITHUB_BASE_REF:-false}")
 DB := example
 IPYTHON := no
@@ -18,24 +17,6 @@ default:
 	| egrep -v -e '^[^[:alnum:]]' -e '^$@$$'
 	@echo
 	@exit 1
-
-.state/env/pyvenv.cfg: requirements/dev.txt requirements/docs.txt requirements/lint.txt requirements/ipython.txt
-	# Create our Python 3.9 virtual environment
-	rm -rf .state/env
-	python3.9 -m venv .state/env
-
-	# install/upgrade general requirements
-	.state/env/bin/python -m pip install --upgrade pip setuptools wheel
-
-	# install various types of requirements
-	.state/env/bin/python -m pip install -r requirements/dev.txt
-	.state/env/bin/python -m pip install -r requirements/docs.txt
-	.state/env/bin/python -m pip install -r requirements/lint.txt
-
-	# install ipython if enabled
-ifeq ($(IPYTHON),"yes")
-	.state/env/bin/python -m pip install -r requirements/ipython.txt
-endif
 
 .state/docker-build: Dockerfile package.json package-lock.json requirements/main.txt requirements/deploy.txt
 	# Build our docker containers for this project.
@@ -60,33 +41,33 @@ debug: .state/docker-build
 
 tests: .state/docker-build
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/tests --postgresql-host db $(T) $(TESTARGS)
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/tests --postgresql-host db $(T) $(TESTARGS)
 
 static_tests: .state/docker-build
 	docker-compose run --rm static env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/static_tests $(T) $(TESTARGS)
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/static_tests $(T) $(TESTARGS)
 
 static_pipeline: .state/docker-build
 	docker-compose run --rm static env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/static_pipeline $(T) $(TESTARGS)
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/static_pipeline $(T) $(TESTARGS)
 
 reformat: .state/docker-build
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/reformat
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/reformat
 
 lint: .state/docker-build
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/lint && bin/static_lint
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/lint && bin/static_lint
 
 docs: .state/docker-build
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/docs
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/docs
 
 licenses:
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
@@ -95,13 +76,13 @@ licenses:
 
 deps: .state/docker-build
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/deps
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/deps
 
 requirements/%.txt: requirements/%.in
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
 
 github-actions-deps:
 ifneq ($(GITHUB_BASE_REF), false)

--- a/Makefile
+++ b/Makefile
@@ -89,7 +89,9 @@ docs: .state/docker-build
 								  bin/docs
 
 licenses:
-	bin/licenses
+	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
+	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+	  bin/licenses
 
 deps: .state/docker-build
 	docker-compose run --rm web env -i ENCODING="C.UTF-8" \

--- a/Makefile
+++ b/Makefile
@@ -40,49 +40,34 @@ debug: .state/docker-build
 	docker-compose run --rm --service-ports web
 
 tests: .state/docker-build
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/tests --postgresql-host db $(T) $(TESTARGS)
+	docker-compose run --rm web bin/tests --postgresql-host db $(T) $(TESTARGS)
 
 static_tests: .state/docker-build
-	docker-compose run --rm static env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/static_tests $(T) $(TESTARGS)
+	docker-compose run --rm static bin/static_tests $(T) $(TESTARGS)
 
 static_pipeline: .state/docker-build
-	docker-compose run --rm static env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/static_pipeline $(T) $(TESTARGS)
+	docker-compose run --rm static bin/static_pipeline $(T) $(TESTARGS)
 
 reformat: .state/docker-build
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/reformat
+	docker-compose run --rm web bin/reformat
 
 lint: .state/docker-build
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/lint && bin/static_lint
+	docker-compose run --rm web bin/lint && bin/static_lint
 
 docs: .state/docker-build
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/docs
+	docker-compose run --rm web bin/docs
 
-licenses:
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/licenses
+licenses: .state/docker-build
+	docker-compose run --rm web bin/licenses
 
 deps: .state/docker-build
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/deps
+	docker-compose run --rm web bin/deps
+
+translations: .state/docker-build
+	docker-compose run --rm web env -i  bin/translations
 
 requirements/%.txt: requirements/%.in
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-	  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-	  bin/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
+	docker-compose run --rm web bin/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
 
 initdb:
 	docker-compose run --rm web psql -h db -d postgres -U postgres -c "SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE datname ='warehouse';"
@@ -108,12 +93,5 @@ purge: stop clean
 
 stop:
 	docker-compose down -v
-
-translations: .state/docker-build
-	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
-								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
-								  bin/translations
-
-
 
 .PHONY: default build serve initdb shell tests docs deps clean purge debug stop compile-pot

--- a/Makefile
+++ b/Makefile
@@ -64,7 +64,7 @@ deps: .state/docker-build
 	docker-compose run --rm web bin/deps
 
 translations: .state/docker-build
-	docker-compose run --rm web env -i  bin/translations
+	docker-compose run --rm web bin/translations
 
 requirements/%.txt: requirements/%.in
 	docker-compose run --rm web bin/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<

--- a/Makefile
+++ b/Makefile
@@ -108,9 +108,10 @@ lint: .state/docker-build
 								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
 								  bin/lint && bin/static_lint
 
-docs: .state/env/pyvenv.cfg
-	$(MAKE) -C docs/ doctest SPHINXOPTS="-W" SPHINXBUILD="$(BINDIR)/sphinx-build"
-	$(MAKE) -C docs/ html SPHINXOPTS="-W" SPHINXBUILD="$(BINDIR)/sphinx-build"
+docs: .state/docker-build
+	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
+								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+								  bin/docs
 
 licenses:
 	bin/licenses

--- a/Makefile
+++ b/Makefile
@@ -96,8 +96,10 @@ deps: .state/docker-build
 								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
 								  bin/deps
 
-requirements/%.txt: requirements/%.in .state/env/pyvenv.cfg
-	$(BINDIR)/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
+requirements/%.txt: requirements/%.in
+	docker-compose run --rm web env -i ENCODING="C.UTF-8" \
+								  PATH="/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \
+								  bin/pip-compile --allow-unsafe --generate-hashes --output-file=$@ $<
 
 github-actions-deps:
 ifneq ($(GITHUB_BASE_REF), false)

--- a/bin/depchecker.py
+++ b/bin/depchecker.py
@@ -1,0 +1,33 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.import sys
+
+import sys
+
+from pip_api import parse_requirements
+
+left, right = sys.argv[1:3]
+left_reqs = parse_requirements(left).keys()
+right_reqs = parse_requirements(right).keys()
+
+extra_in_left = left_reqs - right_reqs
+extra_in_right = right_reqs - left_reqs
+
+if extra_in_left:
+    for dep in sorted(extra_in_left):
+        print("- {}".format(dep))
+
+if extra_in_right:
+    for dep in sorted(extra_in_right):
+        print("+ {}".format(dep))
+
+if extra_in_left or extra_in_right:
+    sys.exit(1)

--- a/bin/deps
+++ b/bin/deps
@@ -7,8 +7,6 @@ set -e
 export LC_ALL="${ENCODING:-en_US.UTF-8}"
 export LANG="${ENCODING:-en_US.UTF-8}"
 
-COMMAND_ARGS=$@
-
 # Print all the followng commands
 set -x
 

--- a/bin/deps
+++ b/bin/deps
@@ -1,0 +1,23 @@
+#!/bin/bash
+set -e
+
+# Click requires us to ensure we have a well configured environment to run
+# our click commands. So we'll set our environment to ensure our locale is
+# correct.
+export LC_ALL="${ENCODING:-en_US.UTF-8}"
+export LANG="${ENCODING:-en_US.UTF-8}"
+
+COMMAND_ARGS=$@
+
+# Print all the followng commands
+set -x
+
+export TMPDIR=$(mktemp -d)
+pip-compile --upgrade --allow-unsafe -o $TMPDIR/deploy.txt requirements/deploy.in > /dev/null
+pip-compile --upgrade --allow-unsafe -o $TMPDIR/main.txt requirements/main.in > /dev/null
+pip-compile --upgrade --allow-unsafe -o $TMPDIR/lint.txt requirements/lint.in > /dev/null
+python bin/depchecker.py $TMPDIR/deploy.txt requirements/deploy.txt
+python bin/depchecker.py $TMPDIR/main.txt requirements/main.txt
+python bin/depchecker.py $TMPDIR/lint.txt requirements/lint.txt
+rm -r $TMPDIR
+pip check

--- a/bin/docs
+++ b/bin/docs
@@ -1,0 +1,16 @@
+#!/bin/bash
+set -e
+
+# Click requires us to ensure we have a well configured environment to run
+# our click commands. So we'll set our environment to ensure our locale is
+# correct.
+export LC_ALL="${ENCODING:-en_US.UTF-8}"
+export LANG="${ENCODING:-en_US.UTF-8}"
+
+COMMAND_ARGS=$@
+
+# Print all the followng commands
+set -x
+
+make -C docs/ doctest SPHINXOPTS="-W"
+make -C docs/ html SPHINXOPTS="-W"

--- a/bin/docs
+++ b/bin/docs
@@ -7,8 +7,6 @@ set -e
 export LC_ALL="${ENCODING:-en_US.UTF-8}"
 export LANG="${ENCODING:-en_US.UTF-8}"
 
-COMMAND_ARGS=$@
-
 # Print all the followng commands
 set -x
 

--- a/bin/github-actions-deps
+++ b/bin/github-actions-deps
@@ -1,0 +1,19 @@
+#!/bin/bash
+set -e
+
+# Click requires us to ensure we have a well configured environment to run
+# our click commands. So we'll set our environment to ensure our locale is
+# correct.
+export LC_ALL="${ENCODING:-en_US.UTF-8}"
+export LANG="${ENCODING:-en_US.UTF-8}"
+
+# Print all the followng commands
+set -x
+
+if [[ -z "${GITHUB_BASE_REF}" ]]; then
+    git fetch origin $(GITHUB_BASE_REF):refs/remotes/origin/$(GITHUB_BASE_REF)
+    # Check that the following diff will exit with 0 or 1
+    git diff --name-only FETCH_HEAD || test $? -le 1 || exit 1
+    # Make the dependencies if any changed files are requirements files, otherwise exit
+    git diff --name-only FETCH_HEAD | grep '^requirements/' || exit 0 && bin/deps
+fi

--- a/bin/licenses
+++ b/bin/licenses
@@ -10,8 +10,11 @@ for fn in $(find . -type f \( \
           ! -name "*.min.js" \
             -name "*.js" \) \
           ! -path "./.state*" \
-          ! -path "./node_modules*" \
+          ! -path "./docs*" \
           ! -path "./htmlcov*" \
+          ! -path "./node_modules*" \
+          ! -path "./warehouse/admin/static/dist*" \
+          ! -path "./warehouse/static/dist*" \
           ! -path "./warehouse/static/html*" \
           ! -path "./warehouse/static/js/vendor*"); do
     # Check for license in first 5 lines (allows for shebang, encoding and handles comment character in various languages)

--- a/bin/lint
+++ b/bin/lint
@@ -7,8 +7,6 @@ set -e
 export LC_ALL="${ENCODING:-en_US.UTF-8}"
 export LANG="${ENCODING:-en_US.UTF-8}"
 
-COMMAND_ARGS=$@
-
 # Print all the followng commands
 set -x
 

--- a/bin/pip-compile
+++ b/bin/pip-compile
@@ -1,0 +1,15 @@
+#!/bin/bash
+set -e
+
+# Click requires us to ensure we have a well configured environment to run
+# our click commands. So we'll set our environment to ensure our locale is
+# correct.
+export LC_ALL="${ENCODING:-en_US.UTF-8}"
+export LANG="${ENCODING:-en_US.UTF-8}"
+
+COMMAND_ARGS=$@
+
+# Print all the followng commands
+set -x
+
+pip-compile $COMMAND_ARGS

--- a/bin/pip-compile
+++ b/bin/pip-compile
@@ -7,7 +7,7 @@ set -e
 export LC_ALL="${ENCODING:-en_US.UTF-8}"
 export LANG="${ENCODING:-en_US.UTF-8}"
 
-COMMAND_ARGS=$@
+COMMAND_ARGS="$@"
 
 # Print all the followng commands
 set -x

--- a/bin/tests
+++ b/bin/tests
@@ -7,7 +7,7 @@ set -e
 export LC_ALL="${ENCODING:-en_US.UTF-8}"
 export LANG="${ENCODING:-en_US.UTF-8}"
 
-COMMAND_ARGS=$@
+COMMAND_ARGS="$@"
 
 # Test the postgres connection
 while [ $# -gt 0 ]; do

--- a/bin/translations
+++ b/bin/translations
@@ -7,8 +7,6 @@ set -e
 export LC_ALL="${ENCODING:-en_US.UTF-8}"
 export LANG="${ENCODING:-en_US.UTF-8}"
 
-COMMAND_ARGS=$@
-
 # Print all the followng commands
 set -x
 

--- a/bin/translations
+++ b/bin/translations
@@ -1,0 +1,15 @@
+#!/bin/bash
+set -e
+
+# Click requires us to ensure we have a well configured environment to run
+# our click commands. So we'll set our environment to ensure our locale is
+# correct.
+export LC_ALL="${ENCODING:-en_US.UTF-8}"
+export LANG="${ENCODING:-en_US.UTF-8}"
+
+COMMAND_ARGS=$@
+
+# Print all the followng commands
+set -x
+
+make -C warehouse/locale/ translations

--- a/dev/environment
+++ b/dev/environment
@@ -1,3 +1,6 @@
+ENCODING=C.UTF-8
+PATH=/opt/warehouse/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+
 WAREHOUSE_ENV=development
 WAREHOUSE_TOKEN=insecuretoken
 

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -88,6 +88,7 @@ services:
       - sponsorlogos:/var/opt/warehouse/sponsorlogos
       - simple:/var/opt/warehouse/simple
       - ./bin:/opt/warehouse/src/bin:z
+      - ./requirements:/opt/warehouse/src/requirements:z
     ports:
       - "${WEB_PORT:-80}:8000"
     depends_on:

--- a/docs/development/getting-started.rst
+++ b/docs/development/getting-started.rst
@@ -58,9 +58,9 @@ Configure the development environment
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 .. note::
-   In case you are used to using a venv/virtualenv or virtual environment for Python development:
-   don't use one for warehouse development. Our Makefile scripts and Docker container development flow
-   creates and removes virtualenvs as needed while you are building and testing your work locally.
+   In case you are used to using a virtual environment for Python development:
+   it's unnecessary for Warehouse development. Our Makefile scripts execute all
+   developer actions inside Docker containers.
 
 Why Docker?
 ~~~~~~~~~~~
@@ -68,10 +68,10 @@ Why Docker?
 Docker simplifies development environment set up.
 
 Warehouse uses Docker and `Docker Compose <https://docs.docker.com/compose/>`_
-to automate setting up a "batteries included" development environment.
-The Dockerfile and :file:`docker-compose.yml` files include all the required steps
-for installing and configuring all the required external services of the
-development environment.
+to automate setting up a "batteries included" development environment. The
+:file:`Dockerfile` and :file:`docker-compose.yml` files include all the
+required steps for installing and configuring all the required external
+services of the development environment.
 
 
 Installing Docker
@@ -157,12 +157,6 @@ This will pull down all of the required docker containers, build Warehouse and
 run all of the needed services. The Warehouse repository will be mounted inside
 the Docker container at :file:`/opt/warehouse/src/`. After the initial build,
 you should not have to run this command again.
-
-.. note::
-
-   On some systems ``make build`` may require the prior installation of libpq
-   and its headers so that psycopg2 can be compiled and installed.
-
 
 .. _running-warehouse-containers:
 
@@ -344,21 +338,6 @@ Errors when executing ``make build``
 
       snap install docker
 
-* If you receive the error: ``python3.9: command not found``, ensure you have
-  Python 3.9 installed on your system.
-  This is the "base" Python version that Warehouse uses to create the rest of
-  the development environment.
-
-* If you are using macOS and you see an
-  ``Error: pg_config executable not found.`` message while building ``psycopg``,
-  you can fix it by installing ``libpq`` via ``brew`` and adding it to your
-  ``PATH``:
-
-  .. code-block:: console
-
-      brew install libpq
-      export PATH="/usr/local/opt/libpq/bin:$PATH"
-
 Errors when executing ``make serve``
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
@@ -425,6 +404,9 @@ https://github.com/chadoe/docker-cleanup-volumes)
 This typically occur when Docker is not allocated enough memory to perform the
 migrations. Try modifying your Docker configuration to allow more RAM for each
 container, temporarily stop ``make_serve`` and run ``make initdb`` again.
+
+This may also be due to enabling Compose V2 (see
+https://github.com/pypa/warehouse/issues/10772 for more details).
 
 
 ``make initdb`` complains about PostgreSQL Version

--- a/warehouse/locale/Makefile
+++ b/warehouse/locale/Makefile
@@ -31,7 +31,7 @@ compile-po:
 
 build-mos: compile-pot
 	cd ../../; \
-	$(eval LOCALES := $(shell python -c "from warehouse.i18n import KNOWN_LOCALES; print(' '.join(set(KNOWN_LOCALES)-{'en'}))"))
+	$(eval LOCALES := $(shell python -c "from warehouse.i18n import KNOWN_LOCALES; print(' '.join(set(KNOWN_LOCALES)-{'en'}))")) ; \
 	for LOCALE in $(LOCALES) ; do \
 		L=$$LOCALE $(MAKE) compile-po ; \
 		done

--- a/warehouse/locale/Makefile
+++ b/warehouse/locale/Makefile
@@ -1,0 +1,46 @@
+GITHUB_ACTIONS := $(shell echo "$${GITHUB_ACTIONS:-false}")
+
+compile-pot:
+	cd ../../; \
+	PYTHONPATH=$(PWD) pybabel extract \
+		-F babel.cfg \
+		--omit-header \
+		--output="warehouse/locale/messages.pot" \
+		warehouse
+
+init-po:
+	cd ../../; \
+	pybabel init \
+		--input-file="warehouse/locale/messages.pot" \
+		--output-dir="warehouse/locale/" \
+		--locale="$(L)"
+
+update-po:
+	cd ../../; \
+	pybabel update \
+		--input-file="warehouse/locale/messages.pot" \
+		--output-file="warehouse/locale/$(L)/LC_MESSAGES/messages.po" \
+		--locale="$(L)"
+
+compile-po:
+	cd ../../; \
+	pybabel compile \
+		--input-file="warehouse/locale/$(L)/LC_MESSAGES/messages.po" \
+		--directory="warehouse/locale/" \
+		--locale="$(L)"
+
+build-mos: compile-pot
+	cd ../../; \
+	$(eval LOCALES := $(shell python -c "from warehouse.i18n import KNOWN_LOCALES; print(' '.join(set(KNOWN_LOCALES)-{'en'}))"))
+	for LOCALE in $(LOCALES) ; do \
+		L=$$LOCALE $(MAKE) compile-po ; \
+		done
+
+translations: compile-pot
+ifneq ($(GITHUB_ACTIONS), false)
+	cd ../../; \
+	git diff --quiet ./warehouse/locale/messages.pot || (echo "There are outstanding translations, run 'make translations' and commit the changes."; exit 1)
+else
+endif
+
+.PHONY: compile-pot

